src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs

```diff
@@ -21,62 +24,127 @@ namespace Microsoft.CodeAnalysis.Completion.Providers
     /// </summary>
     internal abstract class AbstractAwaitCompletionProvider : LSPCompletionProvider
     {
+        #region CompletionItem properties keys
+        private const string AwaitCompletionTargetTokenPosition = nameof(AwaitCompletionTargetTokenPosition);
+
+        private static class AwaitCompletionChange
+        {
+            public const string AddAwaitAtCursor = nameof(AwaitCompletionChange) + "." + nameof(AddAwaitAtCursor);
+            public const string AddAwaitBeforeDotExpression = nameof(AwaitCompletionChange) + "." + nameof(AddAwaitBeforeDotExpression);
+            public const string AppendConfigureAwait = nameof(AwaitCompletionChange) + "." + nameof(AppendConfigureAwait);
+            public const string MakeContainerAsync = nameof(AwaitCompletionChange) + "." + nameof(MakeContainerAsync);
+        }
+        #endregion
+
+        protected enum DotAwaitContext
+        {
+            None,
+            AwaitOnly,
+            AwaitAndConfigureAwait,
+        }
+
         /// <summary>
         /// Gets the span start where async keyword should go.
         /// </summary>
         private protected abstract int GetSpanStart(SyntaxNode declaration);
 
         private protected abstract SyntaxNode? GetAsyncSupportingDeclaration(SyntaxToken token);
 
+        private protected abstract ITypeSymbol? GetTypeSymbolOfExpression(SemanticModel semanticModel, SyntaxNode potentialAwaitableExpression, CancellationToken cancellationToken);
+        private protected abstract SyntaxNode? GetExpressionToPlaceAwaitInFrontOf(SyntaxTree syntaxTree, int position, CancellationToken cancellationToken);
+        private protected abstract SyntaxToken? GetDotTokenLeftOfPosition(SyntaxTree syntaxTree, int position, CancellationToken cancellationToken);
```

Comment:
not sure why these are private protected.

Reply:
Maybe @Youssef1313 can tell. He introduced `private protected` in the provider.

Reply:
Change to protected in b5648160